### DISCUSSION TITLE
API Controller to create, update or delete a `User` with `CustomUserSettings`, and assign a `Role`

### DISCUSSION THREAD

**Discussion by MarGraz** [OrchardCMS/OrchardCore#16245] [Q&A]
Hi Everyone,

I'm trying to create an `ApiController` to **create, read, update, or delete** users. This controller will be used in my Orchard Core microservice that will act as an SSO (Single Sign-On) platform. I've already configured the `OpenId` module and set up the `Client Credentials Flow` to call my new SSO ApiController in a secure way.

Now, I took inspiration from [this discussion](https://github.com/OrchardCMS/OrchardCore/discussions/15014#discussioncomment-8065278) to **create a `User` that also has a Content Type** called `UserDetails`, which works with the stereotype `CustomUserSettings` and contains extra information. In my ApiController, I want to create this user with the custom settings, and additionally, I want to assign a `Role` that already exists in Orchard Core, which I will receive in the payload.

I saw that the `UserManager` **should be enough to create the user and assign a role, but I want to ensure I'm doing it right.**

I have created a handler, and the create user API endpoint will call this handler method, you can see it below.

**Questions:**

1. What do you think about the code I wrote?
2. If I set `EmailConfirmed` to false, will the user automatically receive an email to confirm their email address? Or do I need to create a `Workflow` to verify the user email?

```csharp
 public class UserHandler : IUserHandler
{
    private readonly IContentManager _contentManager;
    private readonly UserManager<IUser> _userManager;
    private readonly ILogger<UserHandler> _logger;

    public UserHandler(IContentManager contentManager, UserManager<IUser> userManager, ILogger<UserHandler> logger)
    {
        _contentManager = contentManager;
        _userManager = userManager;
        _logger = logger;
    }

    public async Task<UserApiModel> CreateUserAsync(UserApiModel userApiModel)
    {
        // Check if a user with the provided email already exists
        IUser? existingUser = await _userManager.FindByEmailAsync(userApiModel.Email);
        if (existingUser != null)
        {
            throw new InvalidOperationException("User with this email already exists.");
        }

        // Ensure there is at least one role specified
        if (userApiModel.RoleNames == null || !userApiModel.RoleNames.Any())
        {
            throw new InvalidOperationException("At least one role must be specified.");
        }

        // Create a new user
        User user = new User
        {
            UserName = userApiModel.UserName,
            Email = userApiModel.Email,
            EmailConfirmed = true, // If set to false?
            IsEnabled = userApiModel.IsEnabled,
            TwoFactorEnabled = userApiModel.TwoFactorEnabled,
            IsLockoutEnabled = userApiModel.IsLockoutEnabled
        };

        // Create the user in the UserManager
        IdentityResult userCreateResult = await _userManager.CreateAsync(user);
        if (!userCreateResult.Succeeded)
        {
            throw new InvalidOperationException("Failed to create user.");
        }

        // Assign roles to the user
        IdentityResult roleResult = await _userManager.AddToRolesAsync(user, userApiModel.RoleNames);
        if (!roleResult.Succeeded)
        {
            // Fail rollback
            await _userManager.DeleteAsync(user);
            throw new InvalidOperationException($"Failed to add user to roles: {string.Join(", ", roleResult.Errors.Select(e => e.Description))}");
        }

        // Create a new content item for the user details
        ContentItem userDetailsContentItem = await _contentManager.NewAsync("UserDetails");
        userDetailsContentItem.Published = true;
        userDetailsContentItem.Latest = true;
        userDetailsContentItem.PublishedUtc = DateTime.UtcNow;
        userDetailsContentItem.ModifiedUtc = DateTime.UtcNow;
        userDetailsContentItem.ContentItemVersionId = Guid.NewGuid().ToString();
        
        userDetailsContentItem.Weld("UserDetails", new UserDetails());

        await userDetailsContentItem.AlterAsync<UserDetails>(async userDetailsPart =>
        {
            userDetailsPart.Name = userApiModel.UserDetails.Name;
            userDetailsPart.Surname = userApiModel.UserDetails.Surname;
            userDetailsPart.Phone = userApiModel.UserDetails.Phone;
            userDetailsPart.WhatsApp = userApiModel.UserDetails.WhatsApp;
            userDetailsPart.Street = userApiModel.UserDetails.Street;
            userDetailsPart.StreetNumber = userApiModel.UserDetails.StreetNumber;
            userDetailsPart.City = userApiModel.UserDetails.City;
        });

        // Add the user details content item
        user.Properties.Add("UserDetails", JObject.FromObject(userDetailsContentItem));

        // Update the user with the details information
        IdentityResult updateResult = await _userManager.UpdateAsync(user);
        if (!updateResult.Succeeded)
        {
            throw new InvalidOperationException("Failed to update user with details information.");
        }

        // Populate the ContentItemId and DocumentId from the created user details
            userApiModel.Id = user.Id.ToString();
            userApiModel.UserId = user.UserId;

        return userApiModel;
    }           
}
```

Thank you for your help 😊

**Comment by MarGraz** [comment]
**I found the problem I had with the User creation.** I was not correctly representing the `UserDetails` model to be added as a part😅 I compared my model with the one saved in the DB and found the issue.

Now the create operation seems to work. 
**Here are the changes I made:**

- I **created a new** `UserDetails` model that represents the Orchard Core model part;
- I removed some `userDetailsContentItem` properties, such as `Published`, `Latest`, etc., because I saw from the DB model that **these are not set** when you create a new user from the admin panel;
- **It seems that the `Weld` method is not necessary**, so I removed it 🤔

The modified parts of the code are:

```csharp
[...]

// Create a new content item for the user details
ContentItem userDetailsContentItem = await _contentManager.NewAsync("UserDetails");
// userDetailsContentItem.Published = true;
// userDetailsContentItem.Latest = true;
// userDetailsContentItem.PublishedUtc = DateTime.UtcNow;
// userDetailsContentItem.ModifiedUtc = DateTime.UtcNow;
// userDetailsContentItem.ContentItemVersionId = Guid.NewGuid().ToString();

// userDetailsContentItem.Weld("UserDetails", new UserDetails());

await userDetailsContentItem.AlterAsync<UserDetails>(async userDetailsPart =>
{
    userDetailsPart.Name.Text = userApiModel.UserDetails.Name;
    userDetailsPart.Surname.Text = userApiModel.UserDetails.Surname;
    userDetailsPart.WhatsApp.Text = userApiModel.UserDetails.WhatsApp;
    userDetailsPart.Street.Text = userApiModel.UserDetails.Street;
    userDetailsPart.StreetNumber.Text = userApiModel.UserDetails.StreetNumber;
    userDetailsPart.City.Text = userApiModel.UserDetails.City;
});

[...]
```

This is how the `UserDetails` model saved in the DB looks like:

```json
  "UserDetails": {
    "DisplayText": " ",
    "Name": {
      "Text": "A test"
    },
    "Surname": {
      "Text": "Pluto"
    },
    "Phone": null,
    "WhatsApp": {
      "Text": "+390123564789"
    },
    "Street": {
      "Text": "Via Marconi"
    },
    "StreetNumber": {
      "Text": "10"
    },
    "City": {
      "Text": "Roma"
    },
    "ZipCode": {
      "Text": "00100"
    }
  }
```

**Questions:**

1. What do you think about the code I wrote?
2. If I set `EmailConfirmed` to false, will the user automatically receive an email to confirm their email address? Or do I need to create a `Workflow` to verify the user's email?
3. What are the best practices for deleting and updating a user?

Thank you

**Reply by Piedone** [reply]
1. If it works, then good :).
2. It won't send an e-mail automatically, you'll need to send it. While you can use a workflow for that, I recommend doing it from code. Check out `RegistrationController.RegisterPOST`, then from there follow up to `RegisterUser()` to see how this happens for normal registrations; you can do the same.
3. Best practice around what exactly, do you mean? I recommend you to check out the what the Users module does.

**Reply by Piedone** [reply]
3. Ah OK. Nothing specific comes to my mind.

**Reply by MarGraz** [reply]
Hi @Piedone,

Thank you for your reply, as always 😊

1. Yes, it works;
2. Ok, I will check the `RegistrationController.RegisterPOST` to implement the email send;
3. With best practice, I mean if I need to pay attention to something specific. For example, during my tests, I saw that **it is not necessary to save or publish my `UserDetails` content item** (`UserDetails` is my Content Type, that has the `CustomUserSettings` stereotype to add more info to the user profile).

I saw that **saving/publishing it will create a zombie Content Item in the DB**, because when I delete the user, it is not deleted accordingly. Basically, it is not necessary to save/publish a `CustomUserSettings` because it will be saved inside the `Document` table, in the user object, inside the `Properties` field, as shown in the JSON below.

```json
{
  "Id": 201,
  "UserId": "***pxj",
  "UserName": "LibreriaTest",
  "NormalizedUserName": "LIBRERIATEST",
  "Email": "test@test.com",
  "PhoneNumber": "+3906123456789",
  "NormalizedEmail": "TEST@TEST.COM",
  "SecurityStamp": "***G5D",
  "EmailConfirmed": true,
  "PhoneNumberConfirmed": true,
  "IsEnabled": false,
  "TwoFactorEnabled": false,
  "IsLockoutEnabled": true,
  "AccessFailedCount": 0,
  "RoleNames": [
    "Standard User"
  ],
  "UserClaims": [],
  "LoginInfos": [],
  "UserTokens": [],
  "Properties": {
    "UserDetails": {
      "ContentItemId": "***wn0",
      "ContentItemVersionId": null,
      "ContentType": "UserDetails",
      "DisplayText": null,
      "Latest": false,
      "Published": false,
      "ModifiedUtc": null,
      "PublishedUtc": null,
      "CreatedUtc": null,
      "Owner": null,
      "Author": null,
      "UserDetails": {
        "Name": {
          "Text": "Libreria"
        },
        "Surname": {
          "Text": "Due"
        },
        "WhatsApp": {
          "Text": null
        },
        "Street": {
          "Text": "Via Milano"
        },
        "StreetNumber": {
          "Text": "1"
        },
        "City": {
          "Text": "Roma"
        },
        "ZipCode": {
          "Text": "00100"
        }
      }
    }
  }
}
```

Thank you

**Reply by MarGraz** [reply]
@Piedone I'm working with the `1.8.3` release, which is different from the `main` branch, as a result, **I can't follow** `RegistrationController.RegisterPOST`, **because I can't find any** `IEmailService` **to use**. 

**But in the 1.8.3 I can take inspiration from the** [SendVerificationEmail](https://github.com/OrchardCMS/OrchardCore/blob/be675b3e1805f3a7b30afff9ceab747719ca9091/src/OrchardCore.Modules/OrchardCore.Users/Controllers/RegistrationController.cs#L153C42-L153C63) method in the `RegistrationController`, which uses an extension class called [ControllerExtensions](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Users/Extensions/ControllerExtensions.cs#L109).

I will let you know if it works 😊

Thank you

**Reply by Piedone** [reply]
Oh yeah, there was some moving around of that code indeed. However, if you're building a new app, I recommend you use the [preview packages](https://docs.orchardcore.net/en/latest/getting-started/preview-package-source/), and then you'll get the very latest OC. This also includes a migration from Json.NET to System.Text.Json, a crucial change that you as a module author also need to adapt to, so right now it's better to start with the preview.

**Reply by MarGraz** [reply]
Hi,

**About my question at point number 2, to send email for user confirmation**, I want to share what I have done in Orchard Core 1.8.3. 
I created a handler class with two methods: `SendEmail` and `SendEmailConfirmationToken`. The `SendEmail` method works as an email sender, and the `SendEmailConfirmationToken` method instead sends the user verification token via email for new user registrations.

Here is the email handler:

```csharp
public class EmailHandler : IEmailHandler
{
    private readonly ISmtpService _smtpService;
    private readonly IDisplayHelper _displayHelper;
    private readonly HtmlEncoder _htmlEncoder;
    private readonly IUrlHelper _urlHelper;
    private readonly UserManager<IUser> _userManager;

    public EmailHandler(ISmtpService smtpService, IDisplayHelper displayHelper, HtmlEncoder htmlEncoder, IUrlHelperFactory urlHelperFactory, IHttpContextAccessor httpContextAccessor, UserManager<IUser> userManager)
    {
        _smtpService = smtpService;
        _displayHelper = displayHelper;
        _htmlEncoder = htmlEncoder;
        _urlHelper = urlHelperFactory.GetUrlHelper(new ActionContext(httpContextAccessor.HttpContext, httpContextAccessor.HttpContext.GetRouteData(), new ActionDescriptor()));
        _userManager = userManager;
    }

    public async Task<bool> SendEmailAsync(string email, string subject, IShape model)
    {
        var body = string.Empty;

        using (var sw = new StringWriter())
        {
            IHtmlContent htmlContent = await _displayHelper.ShapeExecuteAsync(model);
            htmlContent.WriteTo(sw, _htmlEncoder);
            body = sw.ToString();
        }

        MailMessage message = new MailMessage()
        {
            To = email,
            Subject = subject,
            Body = body,
            IsHtmlBody = true
        };

        SmtpResult result = await _smtpService.SendAsync(message);
        return result.Succeeded;
    }

    public async Task<string> SendEmailConfirmationTokenAsync(User user, string subject)
    {
        string code = await _userManager.GenerateEmailConfirmationTokenAsync(user);
        string? callbackUrl = _urlHelper.Action("ConfirmUserEmail", "MyEmailApi", new { userId = user.UserId, code }, protocol: "https");

        await SendEmailAsync(user.Email, subject, new ConfirmEmailViewModel() { User = user, ConfirmEmailUrl = callbackUrl });
        return callbackUrl;
    }
}
```

**What I noticed is that, because I have the feature `UsersRegistration` turned off**, because by our requirements we don't need that the user can register by itself, it was necessary to implement a `Controller` that manage the callback url. This because the default endpoints are not present with the `UsersRegistration` turned off. Infact, this part of code was not returning a callback URL:

```csharp
    public async Task<string> SendEmailConfirmationTokenAsync(User user, string subject)
    {
        [...]

        string? callbackUrl = _urlHelper.Action("ConfirmUserEmail", "MyEmailApi", new { userId = user.UserId, code }, protocol: "https");

        [...]
    }
```

This is the Controller that I have implemented to make it works. Of course I also implemented a method in the UserHandler to be able confirm the user via callback URL:

```csharp
[Authorize(AuthenticationSchemes = "CustomApi"), IgnoreAntiforgeryToken]
[ApiController]
public class MyEmailApiController : Controller
{
 private readonly IUserHandler _userHandler;

 public MyEmailApiController(IUserHandler userHandler)
 {
    _userHandler = userHandler;
 }

 // Confirms the email of a user via callback URL sent by email
 [HttpGet("confirmEmail")]
 public async Task<IActionResult> ConfirmUserEmailAsync([FromForm] string userId, [FromForm] string code)
 {
   // In the UserHandler is necessary to add a method that will confirm the user in the Orchard Core DB
   var result = await _userHandler.ConfirmUserEmailAsync(userId, code);
   return Ok(result);
 }
}
```

I hope it helps 😊
Thank you

**Reply by Piedone** [reply]
Thanks for sharing this!

**Reply by MarGraz** [reply]
> Oh yeah, there was some moving around of that code indeed. However, if you're building a new app, I recommend you use the [preview packages](https://docs.orchardcore.net/en/latest/getting-started/preview-package-source/), and then you'll get the very latest OC. This also includes a migration from Json.NET to System.Text.Json, a crucial change that you as a module author also need to adapt to, so right now it's better to start with the preview.

Hi @Piedone,

I usually use **stable versions** of libraries, CMS, or frameworks **when I want to take a project into production**. 
So, I have few questions:

1. Can Orchard Core `preview` be considered stable, or is it an `alpha` or `beta` that is subject to many changes?
2. What kind of changes should I expect in the `preview` source code?
3. How difficult is it to upgrade from 1.8.3 to the preview now, considering I have already created 7 Orchard Core microservices, various Content Types, and two modules?
4. Does the upgrade involve a radical change in the database requiring manual intervention and a lot of effort, or is the migration automatic and not radical?

I'm asking these questions to evaluate the impact on my effort before making the change. Otherwise, I prefer to postpone this upgrade to a later time 🤔

Thank you

**Reply by Piedone** [reply]
1. The preview is not meant to be stable, but in practice currently it fairly is.
2. Anything that the community works on and merges in the meantime from the currently open, and later opened pull requests. This can include breaking changes, since the upcoming version will be 2.0.
3. See the 2.0 release notes: https://docs.orchardcore.net/en/latest/releases/2.0.0/ These are the upgrade steps from 1.8.x to the current preview. If there will be anything else for the actually released 2.0, then that'll be added there.
4. Also see the release notes. The migration to STJ involves manual updates to your code, with potentially (depending on your code) in-depth changes.

**Reply by MarGraz** [reply]
> 1. The preview is not meant to be stable, but in practice currently it fairly is.

That's good.

> 2. Anything that the community works on and merges in the meantime from the currently open, and later opened pull requests. This can include breaking changes, since the upcoming version will be 2.0.

That's not good considering that I will be in a production environment next month.

> 3. See the 2.0 release notes: https://docs.orchardcore.net/en/latest/releases/2.0.0/ These are the upgrade steps from 1.8.x to the current preview. If there will be anything else for the actually released 2.0, then that'll be added there.

Nice, all the changes are well documented 😊

> 4. Also see the release notes. The migration to STJ involves manual updates to your code, with potentially (depending on your code) in-depth changes.

Ok, I think that because I need to go into production next month, **it's better to use a conservative strategy**. I will continue developing on 1.8.3, and when 2.0 is released, I will manage the update with a calculated effort.

Thank you for all the explanations, as always 🙃

**Reply by Piedone** [reply]
No problem!
